spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/filter/factory/RequestHeaderSizeGatewayFilterFactory.java

```diff
@@ -0,0 +1,98 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.springframework.cloud.gateway.filter.factory;
+
+import org.springframework.cloud.gateway.filter.GatewayFilter;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.server.reactive.ServerHttpRequest;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This filter validates the size of each Request Header in the request.
+ * If size of any of the request header is greater than the configured maxSize,it blocks the request.
+ * Default max size of request header is 16KB.
+ * @author Sakalya Deshpande
+ */
+
+public class RequestHeaderSizeGatewayFilterFactory
+		extends AbstractGatewayFilterFactory<RequestHeaderSizeGatewayFilterFactory.Config> {
+
+	private static String PREFIX = "kMGTPE";
+	private static String ERROR = "Request Header/s size is larger than permissible limit."
+			+ " Request Header/s size is %s where permissible limit is %s";
+
+
+	public RequestHeaderSizeGatewayFilterFactory() {
+		super(RequestHeaderSizeGatewayFilterFactory.Config.class);
+	}
+
+	@Override
+	public GatewayFilter apply(RequestHeaderSizeGatewayFilterFactory.Config config) {
+		return (exchange, chain) -> {
+			ServerHttpRequest request = exchange.getRequest();
+			HttpHeaders headers = request.getHeaders();
+			Long headerSizeInBytes = 0L;
+
+			for(Map.Entry<String,List<String>> headerEntry : headers.entrySet()){
+				List<String> values = headerEntry.getValue();
+				for(String value : values){
+					headerSizeInBytes += Long.valueOf(value.getBytes().length);
+				}
+			}
+
+			if(headerSizeInBytes > config.getMaxSize()){
+				exchange.getResponse().setStatusCode(HttpStatus.REQUEST_HEADER_FIELDS_TOO_LARGE);
+				exchange.getResponse().getHeaders().add("errorMessage",
+						getErrorMessage(headerSizeInBytes, config.getMaxSize()));
+				return exchange.getResponse().setComplete();
+
+			}
+
+			return chain.filter(exchange);
+		};
+	}
+
+	private static String getErrorMessage(Long currentRequestSize, Long maxSize) {
+		return String.format(ERROR, getReadableByteCount(currentRequestSize), getReadableByteCount(maxSize));
+	}
+
+	private static String getReadableByteCount(long bytes) {
+		int unit = 1000;
+		if (bytes < unit)
+			return bytes + " B";
+		int exp = (int) (Math.log(bytes) / Math.log(unit));
+		String pre = Character.toString(PREFIX.charAt(exp - 1));
+		return String.format("%.1f %sB", bytes / Math.pow(unit, exp), pre);
+	}
+
+	public static class Config {
+		private Long maxSize = 16000L;
+
+		public Long getMaxSize() {
```

Comment:
Spring has a [`DataSize`](https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/util/unit/DataSize.html) class that should be used instead of `Long`. It will handle all conversions to/from strings and to bytes.

Reply:
Made change accordingly